library/ssl_tls.c

```diff
@@ -3988,10 +3988,52 @@ void mbedtls_ssl_conf_dhm_min_bitlen( mbedtls_ssl_config *conf,
 /*
  * Set allowed/preferred hashes for handshake signatures
  */
-void mbedtls_ssl_conf_sig_hashes( mbedtls_ssl_config *conf,
-                                  const int *hashes )
+void mbedtls_ssl_conf_sig_hashes( mbedtls_ssl_config *conf, const int *hashes )
 {
-    conf->sig_hashes = hashes;
+    const int *md;
+    size_t sig_algs_len = 0;
+    uint16_t *p;
+
+    if( conf->sig_algs != NULL )
+    {
+        mbedtls_free( ( void * ) conf->sig_algs );
+        conf->sig_algs = NULL;
+    }
+
+    for( md = hashes; *md != MBEDTLS_MD_NONE; md++ )
+    {
+        if( mbedtls_ssl_hash_from_md_alg( *md ) == MBEDTLS_SSL_HASH_NONE )
+            continue;
+#if defined(MBEDTLS_ECDSA_C)
+        sig_algs_len++;
+#endif
+#if defined(MBEDTLS_RSA_C)
+        sig_algs_len++;
+#endif
+    }
+    conf->sig_algs = mbedtls_calloc( sig_algs_len + 1, sizeof( uint16_t ) );
+    if( conf->sig_algs == NULL )
```

Comment:
This is problematic but we have been through the same kind of problem, I think, with the move from curves to groups, see `mbedtls_ssl_conf_curves()` and `mbedtls_ssl_conf_groups()`. In the present case curves are hashes and groups are signature algorithms. Thus it would be probably better to adopt the same solution here and we would keep `mbedtls_ssl_conf_hashes()` but deprecate it and add `mbedtls_ssl_conf_sig_algs()` to be used both in TLS 1.2 and TLS 1.3 case. If `mbedtls_ssl_conf_hashes()` is used to configure then the translation to signature algorithms would be done in `ssl_handshake_init()`, see the section of code starting at "if ( ssl->conf->curve_list != NULL ) ".